Promiscuous mode support

jukkar opened the pull request.

jukkar committed:
net: if: Add promiscuous mode set / unset functionality

User is able to set the network interface to promiscuous mode
and query the promisc mode status.

Note that currently this is only supported for ethernet bearer.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: eth: Add generic promiscuous mode support

Allow ethernet L2 driver to set / unset the device driver
promiscuous mode flag.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
tests: net: ethernet_mgmt: Add promisc mode tests

Test in ethernet mgmt level that promiscuous flag is passed
properly to the device driver.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
tests: net: if: Add promisc mode tests

Make sure that user is able to toggle promiscuous mode flag
in network interface level and that flag value is passed
to device driver.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: eth: native_posix: Return proper error code from linux

Make sure that the system() call will return proper error code
to the ethernet driver.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: eth: native_posix: Add promiscuous mode support

Allow the zeth network interface to be placed into promiscuous mode.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: Add promiscuous mode support

Allow user to set the network interface into promiscuous mode
and then receive all the network packets that are received by
that interface.

Fixes #7595

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: shell: Print information about promiscuous mode

If the network interface is in promiscuous mode, print information
about it.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: Add promiscuous mode application

Simple application that sets all the network interfaces into
promiscuous mode.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

tbursztyka commented on subsys/net/ip/promiscuous.c:
```diff
@@ -0,0 +1,88 @@
+		return -EINVAL;
+	}
+
+	if (!atomic_cas(&enabled, 0, 1)) {
```
ok works on one iface at a time. Fine by me, let's see if we get the use case for 2+ iface at some point

jukkar replied:
Hmm, you are right. Actually it was meant to work with multiple interfaces as seen by the sample app. Let's see if it would be easy to add that support.

tbursztyka reviewed (approved):
Looks fine

jukkar committed:
tests: net: Add tests for net_promisc API

Test that user is able to turn on/off promiscuous mode and
also receive any data from network interface.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
* Added multi interface support
* Added unit tests for the net_promisc_* API

pfalcon commented on subsys/net/ip/net_if.c:
```diff
@@ -2151,6 +2151,47 @@ struct net_if_mcast_addr *net_if_ipv4_maddr_lookup(const struct in_addr *maddr,
+		 * after it has finished working with it.
+		 *
+		 * If packet is for us, then NET_CONTINUE is returned.
+		 * In this case we must clone the packet, as the packet
```
Why all these complications? Why not call promisc handler first, and make a stipulation that it should not alter packet (if it needs to, it should clone it and operate on clone).

jukkar replied:
There is a pipe between core stack and application for these promiscuous packets which means that packet might be destroyed already when application gets it. So in order to avoid any issue, it is cloned here.

pfalcon reviewed (dismissed):
Two notes:

1. Seeing k_fifo here, it immediately leads to thought: "poorman's socket work-alike". It's similar in this respect to initial patch for TLS, which just duplicated good deal of existing socket processing. Ok, I won't call for adding SOCK_RAW support and putting all this stuff under sockets, but we all should anticipate that in the future net_promisc_mode_wait_data() will be gone completely and will become recv() on a raw socket instead. (That should be ok, as the current API is marked EXPERIMENTAL).

2. And the reason why I give -1 to this. All the net_if_recv_data() hackery should be unnecessary. It should just pass pkt to net_promisc_mode_input(), and *it*, if needed, should make a pkt clone, and put that in the fifo.

jukkar commented:
> All the net_if_recv_data() hackery should be unnecessary. It should just pass pkt to net_promisc_mode_input(), and it, if needed, should make a pkt clone, and put that in the fifo.

There is no hackery involved and what you suggest is not possible. We need to know if the packet is meant for "us" or is it for some other device. For this information we need to call corresponding L2 recv function which returns this info. It would make no sense to try to put this functionality into net_promisc_mode_input().

pfalcon commented:
> We need to know if the packet is meant for "us" or is it for some other device. For this information we need to call corresponding L2 recv function which returns this info.

Then there should be code comment that what's been tried there is to optimize cloning, i.e. if packet is "not for us", then we can avoid cloning, otherwise we need 2 copies - one for local processing, and one for promisc queue.

jukkar commented:
> Then there should be code comment that what's been tried there is to optimize cloning, i.e. if packet is "not for us", then we can avoid cloning, otherwise we need 2 copies - one for local processing, and one for promisc queue.

There is already a comment in the code about this. Currently it says
`If the packet is not for us and the promiscuous mode is enabled, then increase the ref count so that net_core.c:processing_data() will not free it. The promiscuous mode handler must free the packet after it has finished working with it. If packet is for us, then NET_CONTINUE is returned. In this case we must clone the packet, as the packet could be manipulated by other part of the stack.
`
